### ISSUE TITLE
cosmic-greeter doesn't remember keyboard layout after reboot

### ISSUE THREAD

**Issue by Thatoo** [pop-os/cosmic-epoch#1160]
If I'm logged in and my computer goes to sleep, then cosmic-greeter  remember my keyboard layout.

However, after a reboot, cosmic-greeter doesn't remember my keyboard layout and I have to deal with US layout to type my password.

**Comment by Thatoo** [comment]
This, with numlock are the two most daily annoying issue I face because typing password with wrong keyboard layout (being US keyboard or my keyboard layout but without numlock) makes me type wrong password and I shoudl type again and again...

I'd really appreciate if cosmic greeter could remember the keyboard layout (with or without numlock) that was set when the user logout/switch off after a reboot (and after sleep of course).

**Comment by joshuamegnauth54** [comment]
What happens when you change your keyboard layout within COSMIC Greeter itself?

I'm trying to tackle this but I only have one layout so it's hard to test. :pensive:

**Comment by Thatoo** [comment]
Well, the thing is that even though, I have set two layouts in cosmic settings :
 
![image](https://github.com/user-attachments/assets/187c3abd-d948-448b-b659-cfc697ae17a7)

I don't have any input sources applet showing within cosmic (should appear between sond and power applets) : 

![image](https://github.com/user-attachments/assets/d9483aa0-79fb-4cb0-a698-023fe98afec9)

and when I'm in cosmic greeter, I can't choose (change input sources) : 

![IMG_20241115_113744](https://github.com/user-attachments/assets/d0048f00-7643-48de-b1a4-4050cdde188b)

**Comment by joshuamegnauth54** [comment]
Thanks for your screenshots!

I am able to replicate this by deleting my user's COSMIC Comp's `xkb_config`, but it is correctly recreated when I login again.

Can you check if `~/.config/cosmic/com.system76.CosmicComp/v1/xkb_config` and copy its contents here?

Also, check: `/var/lib/cosmic-greeter/.config/com.system76.CosmicComp/v1/xkb_config`. You may need to do:

```sh
sudo -u cosmic-greeter -s
cd /var/lib/cosmic-greeter
```

In order to check.

**Comment by Thatoo** [comment]
```
~$ cat ~/.config/cosmic/com.system76.CosmicComp/v1/xkb_config
(
    rules: "",
    model: "pc104",
    layout: "fr,fr",
    variant: ",ergol",
    options: Some("terminate:ctrl_alt_bksp"),
    repeat_delay: 600,
    repeat_rate: 25,
```

`/var/lib/cosmic-greeter/.config/com.system76.CosmicComp/v1/xkb_config` doesn't exist : 

```
)thatoo@ThinkPadT470:~$ cat/var/lib/cosmic-greeter/.config/com.system76.CosmicComp/v1/xkb_configg
cat: /var/lib/cosmic-greeter/.config/com.system76.CosmicComp/v1/xkb_config: Permission denied
thatoo@ThinkPadT470:~$ sudo cat /var/lib/cosmic-greeter/.config/com.system76.CosmicComp/v1/xkb_config
[sudo] Mot de passe de thatoo :
cat: /var/lib/cosmic-greeter/.config/com.system76.CosmicComp/v1/xkb_config: No such file or directory
thatoo@ThinkPadT470:~$ sudo -u cosmic-greeter -s
[cosmic-greeter@ThinkPadT470:/home/thatoo]$ cat /var/lib/cosmic-greeter/.config/com.system76.CosmicComp/v1/xkb_config
cat: /var/lib/cosmic-greeter/.config/com.system76.CosmicComp/v1/xkb_config: No such file or directory

[cosmic-greeter@ThinkPadT470:/home/thatoo]$ cd /var/lib/cosmic-greeter

[cosmic-greeter@ThinkPadT470:~]$ cat .config/com.system76.CosmicComp/v1/xkb_config
cat: .config/com.system76.CosmicComp/v1/xkb_config: No such file or directory

[cosmic-greeter@ThinkPadT470:~]$ ls -al
total 20
drwx------  5 cosmic-greeter cosmic-greeter 4096  1 juin  20:23 .
drwxr-xr-x 31 root           root           4096 16 nov.  10:15 ..
drwx------  4 cosmic-greeter cosmic-greeter 4096  1 nov.  09:15 .cache
drwxr-xr-x  3 cosmic-greeter cosmic-greeter 4096  1 juin  20:23 .config
drwxr-xr-x  4 cosmic-greeter cosmic-greeter 4096 15 nov.  11:37 .local

[cosmic-greeter@ThinkPadT470:~]$ cd .config/

[cosmic-greeter@ThinkPadT470:~/.config]$ ls -al
total 12
drwxr-xr-x  3 cosmic-greeter cosmic-greeter 4096  1 juin  20:23 .
drwx------  5 cosmic-greeter cosmic-greeter 4096  1 juin  20:23 ..
drwxr-xr-x 10 cosmic-greeter cosmic-greeter 4096  9 sept. 14:56 cosmic

```

**Comment by joshuamegnauth54** [comment]
Oh wait :sweat_smile: I messed up the path. It should be: `/var/lib/cosmic-greeter/.config/cosmic/com.system76.CosmicComp/v1/xkb_config`

If it's still missing then I think I know the problem and how to fix it.

**Comment by Thatoo** [comment]
```
[cosmic-greeter@mycomputer]$ ls -al /var/lib/cosmic-greeter/.config/cosmic/com.system76.CosmicComp/v1/
total 8
drwxr-xr-x 2 cosmic-greeter cosmic-greeter 4096  1 juin  20:23 .
drwxr-xr-x 3 cosmic-greeter cosmic-greeter 4096  1 juin  20:23 ..
```

Indeed, at the path you indicate, `xkb_config` file doesn't exist.

**Comment by joshuamegnauth54** [comment]
Thanks! I think I have an idea of how to fix it, and I'm assuming you can guess the fix too. :grin:

I'll start working on it today.

**Comment by Thatoo** [comment]
Indeed, `sudo cp ~/.config/cosmic/com.system76.CosmicComp/v1/xkb_config /var/lib/cosmic-greeter/.config/cosmic/com.system76.CosmicComp/v1/xkb_config` solve the issue.

However, at login, I still can't choose (change input sources) :

![IMG_20241115_113744](https://github.com/user-attachments/assets/d0048f00-7643-48de-b1a4-4050cdde188b)

and input applet still isn't showing (but this is the subject of [an other issue](https://github.com/pop-os/cosmic-epoch/issues/1216))

**Comment by joshuamegnauth54** [comment]
I made some progress on this today but then I got stuck. :joy_cat: It seems like COSMIC Settings is doing the right thing whereas the Input applet and Greeter are not for some reason. The Input applet and Greeter also share the same code for retrieving the layouts, so the fact that they're both producing incorrect results seems like a good hint towards the solution!

On my machine, Greeter recreates `xkb_config` from my user config if I delete it and restart. I'm not sure why it's not doing so for you, but *that* should most likely be the solution.

I'll continue trying tomorrow. We're definitely close. :triumph:

**Comment by andyp123** [comment]
Hi. Sorry to add noise, but I'm also having this issue in the Cosmic Beta, and although I was not using cosmic as much during alpha, I didn't experience this issue with any previous alpha.

I have a Thinkpad X13 with Japanese keyboard layout and it resets to US regularly (seems to be in line with this report, as it's usually after restarting), and when it does, not being a terminal wizard, the only way I can restore the correct layout is to remove the Japanese layout in Cosmic Settings and add it back.

By the way, is detecting keyboard layouts through meta data directly from the hardware possible these days? Sometimes I have a US keyboard connected via USB, so would rather that used US layout with the laptop's keyboard remaining in Japanese. Is this possible? Does anyone else think this would be useful assuming it's possible?

**Comment by leviport** [comment]
I believe this was fixed?
